go/plugins/vertexai/embed.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vertexai
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"runtime"
+
+	aiplatform "cloud.google.com/go/aiplatform/apiv1"
+	"cloud.google.com/go/aiplatform/apiv1/aiplatformpb"
+	"github.com/google/genkit/go/ai"
+	"google.golang.org/api/option"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// embedder implements ai.Embedder.
+type embedder struct {
+	projectID string
+	location  string
+	model     string
+	client    *aiplatform.PredictionClient
+}
+
+// EmbedOptions is options for the Vertex AI embedder.
+// Set [ai.EmbedRequest].Options to a value of type &EmbedOptions.
+type EmbedOptions struct {
+	// Document title.
+	Title string `json:"title,omitempty"`
+	// Task type: RETRIEVAL_QUERY, RETRIEVAL_DOCUMENT, and so forth.
+	// See the Vertex AI text embedding docs.
+	TaskType string `json:"task_type,omitempty"`
+}
+
+// Embed converts a document to a multi-dimensional vector.
+// This implements ai.Embedder.
+func (e *embedder) Embed(ctx context.Context, req *ai.EmbedRequest) ([]float32, error) {
+	var title, taskType string
+	if options, _ := req.Options.(*EmbedOptions); options != nil {
+		title = options.Title
+		taskType = options.TaskType
+	}
+
+	endpoint := fmt.Sprintf("projects/%s/locations/%s/publishers/google/models/%s", e.projectID, e.location, e.model)
+
+	instances := make([]*structpb.Value, 0, len(req.Document.Content))
+	for _, part := range req.Document.Content {
+		fields := map[string]any{
+			"content": part.Text(),
+		}
+		if title != "" {
+			fields["title"] = title
```

Comment:
It seems weird that these get set on every part, instead of once somewhere. I guess that's just how the proto wants them?

Reply:
I agree that it is weird, but it is what the API examples do.